### ISSUE TITLE
Using Artisan::call within a job class results in wrong data passed within the system (if queue:work > 1)

### ISSUE THREAD

**Issue by Damian89** [laravel/framework#36085]
- Laravel Version: 8.25.0
- PHP Version: PHP 8.0.1 (cli) (built: Jan 13 2021 08:22:35) ( NTS )
- Database Driver & Version: Server version: 8.0.22-0ubuntu0.20.04.3 (Ubuntu)
- Queue Driver & Version: Redis server v=5.0.7 sha=00000000:0 malloc=jemalloc-5.2.1 bits=64 build=636cde3b5c7a3923

### Description:

When adding several thousands of jobs (with data A) to queue on a server with __multiple queue:work running__ (via supervisor) and then immidiately adding several thousands jobs (with data B) to the same queue, I can see that the first batch of jobs has access to the data of the second batch of jobs. (and vice versa). See example below.

### Steps To Reproduce:

A simple supervisor configuration is used:
https://pastebin.com/rALV2q58

From the frontend I can start a Job called "Requester", the job class is quite simple:
https://pastebin.com/77QBVzB8

As you can see, I basically generate somewhere else a command and then it gets called via Artisan::call. Straight forward. The command itself is basically just:

php artisan requester:scan--target=[ID] --url=[Url] --ua=[User-Agent]

__Command class:__ https://pastebin.com/XRnskGRw

Now imagine I add 10000 different jobs of this kind into queue system (I am using a dedicated redis server). Via logging I can track, that all jobs are queued, and then processed by our server, I see (via logging), that the job class starts and finishes and in between I see the artisan call beeing started and of course also the underlying artisan command. All with the correct Ids and initial input (target ID, url, ua). I simplified the case so no logging functions are visible.

Lets say we add a big batch of jobs which are basically requesting:
```
http://test1.com (with user agent ua1)

http://test2.com (with user agent ua1)

http://test3.com (with user agent ua1)

http://test4.com (with user agent ua1)

http://test5.com (with user agent ua1)

... [10000]
```

Queuesystem starts.

We immidiately add another 10000 jobs, but this time other urls and other user agent string:

```
http://abc1.com (with user agent abc1)

http://abc2.com (with user agent abc1)

http://abc3.com (with user agent abc1)

http://abc4.com (with user agent abc1)

http://abc5.com (with user agent abc1)

... [10000]
```

So we have ~20000 jobs in our queue system. Thats basically our setup (I tried to minimize it as far as possible).

When we observe the traffic on all our 20000 domains the system currently is requesting, we see that after a short amount of time the data gets mixed up:

```
http://test1.com gets requested with user agent: abc1 (instead of ua1)

http://test2.com gets requested with user agent: abc1 (instead of ua1)

http://abc2.com gets requested with user agent: ua1 (instead of abc1)

http://abc3.com gets requested with user agent: ua1 (instead of abc1)

http://abc4.com gets requested with user agent: ua1 (instead of abc1)

... [10000]
```

Thats basically my problem. I have absolutely no idea why this is happening. __All variables are protected.__

I have simple solution:

Instead of using queue:work, I have to switch to queue:listen.

But: https://laravel.com/docs/8.x/queues

```
Alternatively, you may run the queue:listen command. When using the queue:listen command, 
you don't have to manually restart the worker when you want to reload your updated code or
reset the application state; however, this command is significantly less efficient than 
the queue:work command:
```

I benchmarked it and queue:work is __significantly faster__, thats why I would love to use queue:work instead of queue:listen in production.

### Additionally

- when calling the artisan command from the shell, then everything work normally - obviously its just ONE command, and not 20000, so data wont get mixed up
- when switching to queue:listen, everything works fine - but processing jobs is slow as hell
- when using just ONE queue:work process supervised by supervisor, also everything is fine.
- I compared my entire code and classes against the current github version... identically...
- when using local variables (no definitions at the top of the class), it seems to work - but unfortunately the code gets ugly since one has to pass all those local variables in the protected/private functions instead of just using $this->variable
- i have dozens of other jobs and only those make problems where I call an artisan command, when I have the logic in the jobclass itself it looks fine
- this behaviour can be seen in all laravel versions (5-8) and php 7.(2,3,4) & 8.

### Assumptions

- This somehow looks like a bug in laravel when there are a lot of processes ongoing at the same time. In my specific case
 protected array $requestTemplate gets mixed up.
- It is not intendet to start Artisan:call within a job class

I hope you have some ideas, I am at a point where I think this has to be a laravel bug or insufficient documentation.
If needed I can provide access to a production system and show you a real case with real data.

Best,

Damian

**Comment by driesvints** [comment]
This seems like an extremely odd edge case to me. In general I wouldn't recommend to use Artisan like that because it'll spawn so many extra processes on your server.

Can you first please try one of the support channels below? If you can actually identify this as a bug, feel free to report back.

- [Laracasts Forums](https://laracasts.com/discuss)
- [Laravel.io Forums](https://laravel.io/forum)
- [StackOverflow](https://stackoverflow.com/questions/tagged/laravel)
- [Discord](https://discordapp.com/invite/KxwQuKb)
- [Larachat](https://larachat.co)
- [IRC](https://webchat.freenode.net/?nick=laravelnewbie&channels=%23laravel&prompt=1)

Thanks!

**Comment by driesvints** [comment]
What also would help is the most simplest way to reproduce this issue.

**Comment by Damian89** [comment]
Hey @driesvints , I asked at laracasts and stackoverflow - nobody was able to help.

Further analysis showed that its basically how I am using the variables. If I change to local variables within my artisan command class, then everything works fine (even when started from another job class).

What would be the recommended way to implement such think? The idea behind it was initially that we are also using cli - so we had our logic in commands and :

a) had the possibility queue the command from the frontend
or
b) generate the command as a string and use it in cli

Now I am thinking to add the command code into a seperate job and if we need to call it from cli, the command class could contain something like:

```
function handle() {
  (new App\Jobs\SomeJob($this->options()))->handle();
  return 0;
}
```

What do you think?

For the problem, would it be help if I provide a small repo with laravel and some commands/jobs in it, so you are able to reproduce it?

**Comment by driesvints** [comment]
I can't be of much help here sorry as I'm not familiar with such scenarios. I'm afraid that the best way to go about this is to ask someone who has experience with similar situations.

**Comment by Damian89** [comment]
If someone is reading this, and wants to earn some money (of course I am willing to pay) and has some experience in this, hit me! The software I am talking about is quite massive, > 4 years old and works like a charm (with queue:listen). If someone knows someone else who has experience with this large type of app - also hit me. Currently we are processing ~10-20 million jobs per day (ofc on far more then one server).

**Comment by taylorotwell** [comment]
To me it sounds like some shared state in your application - keep in mind that queue:work keeps the same PHP process running as it churns through jobs. Any static state that you set on any class may still be there when the next job runs, etc. That is why your problem disappears with queue:listen. With queue:listen, the entire framework is rebooted in an entirely new PHP process for each job.

**Comment by Damian89** [comment]
It sounds like that, but I have not idea where this kind of shared state may be found - do you have any kind of info/reading sources/pitfalls etc where I can start my research?
Best,
Damian

**Comment by Damian89** [comment]
I "fixed" it. No idea why, but I cant use variables directly which I declared in the top of the class when Aritsan::call is used within a job class - then just data within the command class gets f***** up.

The solution so far is quite simple, e.g. imagine this class:

https://gist.github.com/Damian89/115d27a1aa5f9d91c8aa5cd139795155

There you see I am setting $this->data and using it in a notification. Won't work for me. With a lot of queued jobs, I will get wrong data in my notifications.

Solution (for me):

https://gist.github.com/Damian89/c9c2f866530376eff8b614b0e99e6820

Using:
a) Private (I switched all to private instead of protected, seems more strictly, but just this change wasnt enought)
b) using a getter function $this->getData()

At the end I am glad that it seem to work - but overall I think this is odd. I have no global state of app or whatsoever.
